File: content-service/content-api-core/src/main/scala/org/ekstep/content/util/HttpRestUtil.scala
package org.ekstep.content.util

import com.mashape.unirest.http.{HttpResponse, Unirest}
import org.apache.commons.lang3.StringUtils
import org.ekstep.common.Platform
import org.ekstep.common.enums.TaxonomyErrorCodes
import org.ekstep.common.exception.ServerException
import org.ekstep.commons
import org.ekstep.commons.{Response, ResponseCode}
import org.ekstep.telemetry.logger.TelemetryManager

import scala.collection.JavaConverters._

/**
  * Utility Object to make REST API Calls
  *
  * @author Kumar Gauraw
  */
object HttpRestUtil {

    val EKSTEP_PLATFORM_API_USERID = "System"
    val DEFAULT_CONTENT_TYPE = "application/json"
    val EKSTEP_API_AUTHORIZATION_KEY = "Bearer " + {
        if (Platform.config.hasPath("dialcode.api.authorization")) Platform.config.getString("dialcode.api.authorization") else ""
    }

    Unirest.setDefaultHeader("Content-Type", DEFAULT_CONTENT_TYPE)
    Unirest.setDefaultHeader("Authorization", EKSTEP_API_AUTHORIZATION_KEY)
    Unirest.setDefaultHeader("user-id", EKSTEP_PLATFORM_API_USERID)

    /**
      *
      * @param uri
      * @param requestMap
      * @param headerParam
      * @return
      */
    def post(uri: String, requestMap: Map[String, AnyRef], headerParam: Map[String, String]): Response = {
        TelemetryManager.log("HttpRestUtil:post |  Request Url:" + uri)
        TelemetryManager.log("HttpRestUtil:post |  Request Body:" + requestMap)
        if (null == headerParam) throw new ServerException("ERR_INVALID_HEADER_PARAM", "Header Parameter is Mandatory.")
        if (null == requestMap) throw new ServerException("ERR_INVALID_REQUEST_BODY", "Request Body is Manadatory")
        try {
            //TODO: Check replacement of toJson with JSONUtils or mkString
            val response: HttpResponse[String] = Unirest.post(uri).headers(headerParam.asJava).body(toJson(requestMap)).asString()
            getResponse(response)
        } catch {
            case e: Exception =>
                TelemetryManager.log("HttpRestUtil:post |  Error is :" + e)
                throw new ServerException("ERR_CALL_API", "Something Went Wrong While Making API Call | Error is: " + e.getMessage)
        }
    }

    /**
      *
      * @param urlWithIdentifier
      * @param queryParam
      * @param headerParam
      * @return
      */
    def get(urlWithIdentifier: String, queryParam: String, headerParam: Map[String, String]): Response = {
        if (null == headerParam) throw new ServerException("ERR_INVALID_HEADER_PARAM", "Header Parameter is Mandatory.")
        val req = if (null != queryParam) urlWithIdentifier + queryParam else urlWithIdentifier
        try {
            val response = Unirest.get(req).headers(headerParam.asJava).asString
            getResponse(response)
        } catch {
            case e: Exception =>
                TelemetryManager.log("HttpRestUtil:get |  Error is :" + e)
                throw new ServerException("ERR_CALL_API", "Something Went Wrong While Making API Call | Error is: " + e.getMessage)
        }
    }

    /**
      *
      * @param response
      * @return
      */
    def getResponse(response: HttpResponse[String]): commons.Response = {
        try {
            val resp: Response = {
                if (StringUtils.isNotBlank(response.getBody))
                    JSONUtils.deserialize[Response](response.getBody)
                else Response(null, null, null, null, ResponseCode.SERVER_ERROR.toString, Some(Map()))
            }
            resp
        } catch {
            case e: Exception =>
                TelemetryManager.info("HttpRestUtil | Error Occurred While Parsing Response Body :" + e)
                throw new ServerException(TaxonomyErrorCodes.SYSTEM_ERROR.name, e.getMessage)
        }
    }

    /**
      *
      * @param query
      * @return
      */
    def toJson(query: Any): String = query match {
        case m: Map[String, Any] => s"{${m.map(toJson(_)).mkString(",")}}"
        case t: (String, Any) => s""""${t._1}":${toJson(t._2)}"""
        case ss: Seq[Any] => s"""[${ss.map(toJson(_)).mkString(",")}]"""
        case s: String => s""""$s""""
        case null => "null"
        case _ => query.toString
    }

}

File: platform-modules/common/src/main/java/org/sunbird/common/util/HttpRestUtil.java
package org.sunbird.common.util;

import java.io.UnsupportedEncodingException;
import java.util.Map;
import org.apache.commons.lang3.StringUtils;
import org.sunbird.common.Platform;
import org.sunbird.common.dto.Response;
import org.sunbird.common.enums.TaxonomyErrorCodes;
import org.sunbird.common.exception.ServerException;
import org.sunbird.telemetry.logger.TelemetryManager;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.Unirest;

/**
 * Utility Class to make REST API Calls
 * 
 * @author gauraw
 *
 */
public class HttpRestUtil {

	private static final String EKSTEP_PLATFORM_API_USERID = "System";
	private static final String DEFAULT_CONTENT_TYPE = "application/json";
	private static String EKSTEP_API_AUTHORIZATION_KEY = null;

	private static Gson gsonObj = new Gson();
	private static ObjectMapper objMapper = new ObjectMapper();

	static {
		String key = Platform.config.hasPath("dialcode.api.authorization")
				? Platform.config.getString("dialcode.api.authorization")
				: "";
		EKSTEP_API_AUTHORIZATION_KEY = "Bearer " + key;
		Unirest.setDefaultHeader("Content-Type", DEFAULT_CONTENT_TYPE);
		Unirest.setDefaultHeader("Authorization", EKSTEP_API_AUTHORIZATION_KEY);
		Unirest.setDefaultHeader("user-id", EKSTEP_PLATFORM_API_USERID);
	}

	/**
	 * @param uri
	 * @param requestMap
	 * @param headerParam
	 * @return
	 * @throws Exception
	 */
	public static Response makePostRequest(String uri, Map<String, Object> requestMap, Map<String, String> headerParam)
			throws Exception {
		TelemetryManager.log("HttpRestUtil:makePostRequest |  Request Url:" + uri);
		TelemetryManager.log("HttpRestUtil:makePostRequest |  Request Body:" + requestMap);

		HttpResponse<String> response = null;

		if (null == headerParam)
			throw new ServerException("ERR_INVALID_HEADER_PARAM", "Header Parameter is Mandatory.");

		if (null == requestMap)
			throw new ServerException("ERR_INVALID_REQUEST_BODY", "Request Body is Manadatory");

		try {
			response = Unirest.post(uri).headers(headerParam)
					.body(gsonObj.toJson(requestMap)).asString();
		} catch (Exception e) {
			throw new ServerException("ERR_CALL_API",
					"Something Went Wrong While Making API Call | Error is: " + e.getMessage());
		}

		return getResponse(response);
	}

	/**
	 * @param urlWithIdentifier
	 * @param queryParam
	 * @param headerParam
	 * @return
	 * @throws Exception
	 */
	public static Response makeGetRequest(String urlWithIdentifier, String queryParam, Map<String, String> headerParam)
			throws Exception {
		if (null == headerParam)
			throw new ServerException("ERR_INVALID_HEADER_PARAM", "Header Parameter is Mandatory.");

		String req = (null != queryParam) ? urlWithIdentifier + queryParam : urlWithIdentifier;

		HttpResponse<String> response = Unirest.get(req).headers(headerParam).asString();
		return getResponse(response);
	}

	private static Response getResponse(HttpResponse<String> response) {
		String body = null;
		Response resp = new Response();
		try {
			body = response.getBody();
			if (StringUtils.isNotBlank(body))
				resp = objMapper.readValue(body, Response.class);
		} catch (UnsupportedEncodingException e) {
			TelemetryManager.info("UnsupportedEncodingException:::::"+ e);
			throw new ServerException(TaxonomyErrorCodes.SYSTEM_ERROR.name(), e.getMessage());
		} catch (Exception e) {
			TelemetryManager.info("Exception:::::"+ e);
			throw new ServerException(TaxonomyErrorCodes.SYSTEM_ERROR.name(), e.getMessage());
		}
		return resp;
	}

}
